src/bepasty/tests/test_rest_server.py

```diff
@@ -0,0 +1,666 @@
+#
+# REST api tests
+#
+
+import os
+import base64
+import threading
+import time
+import hashlib
+from requests.auth import _basic_auth_str
+
+import pytest
+
+from ..app import create_app
+from ..config import Config
+from ..constants import FILENAME, TYPE, LOCKED, SIZE, COMPLETE, HASH, \
+    TIMESTAMP_DOWNLOAD, TIMESTAMP_UPLOAD, TIMESTAMP_MAX_LIFE, TRANSACTION_ID
+from ..utils.date_funcs import time_unit_to_sec
+
+UPLOAD_DATA = b"""\
+#!/usr/bin/python3
+
+print('hello, world')
+"""
+
+
+class FakeTime:
+    """Overwrite time.time to control the timestamp that is used in server"""
+
+    def __init__(self, now=None):
+        self.orig = time.time
+        self.now = now
+        self.set_time(self.now)
+
+    def set_time(self, now=None):
+        self.now = now
+        if now and time.time != self.get_time:
+            time.time = self.get_time
+        elif now is None and time.time == self.get_time:
+            time.time = self.orig
+
+    def get_time(self):
+        return self.now
+
+    def __enter__(self):
+        self.set_time(self.now)
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.set_time(None)
+
+
+def wait_background():
+    """Wait background threads"""
+
+    main_thread = threading.current_thread()
+    for t in threading.enumerate():
+        if t is not main_thread:
+            t.join()
+
+
+@pytest.fixture
+def client_fixture(tmp_path):
+    with FakeTime() as faketime:
+        Config.PERMISSIONS = {
+            'admin': 'admin,list,create,read,delete',
+            'full': 'list,create,read,delete',
+            'none': '',
+        }
+        Config.STORAGE_FILESYSTEM_DIRECTORY = str(tmp_path)
+
+        app = create_app()
+        app.config['TESTING'] = True
+
+        with app.test_client() as client:
+            yield app, client, faketime
+
+        wait_background()
+
+
+class TmpConfig:
+    def __init__(self, app, tmp_config={}):
+        self.config = app.config
+        self.orig_config = {}
+        self.tmp_config = tmp_config
+        for k, v in tmp_config.items():
+            self.orig_config[k] = self.config[k]
+
+    def __enter__(self):
+        for k, v in self.tmp_config.items():
+            self.config[k] = v
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        for k, v in self.orig_config.items():
+            self.config[k] = self.orig_config[k]
+
+
+class RestUrl:
```

Comment:
is this really needed or could you also use flask's `url_for` to build URLs?

Reply:
To get /apis/rest/items, we need this for client context. So I think it is still useful RestUrl for easier and reliable way

```python
        with current_app.test_request_context():
            return url_for('bepasty_apis.items')
```

Reply:
if the tests needing a request context would "require" it via a fixture param, guess this RestUrl class could be removed and we could directly use url_for() inside the test function.

Reply:
this test_request_context() is **fake** request context, just to make url_for() works. so I don't want put other code under this context.